providers/rpc/rpc.go

```diff
@@ -0,0 +1,351 @@
+
+// Algorithm is the type for HMAC algorithms.
+type Algorithm string
+
+// Secrets hold per algorithm slice secrets.
+// These secrets will be used to create HMAC signatures.
+type Secrets map[Algorithm][]string
+
+// Signatures hold per algorithm slice of signatures.
+type Signatures map[Algorithm][]string
+
+// Config defines the configuration for sending rpc notifications.
+type Config struct {
+	// ConsumerURL is the URL where an rpc consumer/listener is running
+	// and to which we will send and receive all notifications.
+	ConsumerURL string
+	// Host is the BMC ip address or hostname or identifier.
+	Host string
+	// Logger is the logger to use for logging.
+	Logger logr.Logger
+	// LogNotificationsDisabled determines whether responses from rpc consumer/listeners will be logged or not.
+	LogNotificationsDisabled bool
+	// Opts are the options for the rpc provider.
+	Opts Opts
+
+	// listenerURL is the URL of the rpc consumer/listener.
+	listenerURL *url.URL
+}
+
+type Opts struct {
+	// Request is the options used to create the rpc HTTP request.
+	Request RequestOpts
+	// Signature is the options used for adding an HMAC signature to an HTTP request.
+	Signature SignatureOpts
+	// HMAC is the options used to create a HMAC signature.
+	HMAC HMACOpts
+	// Experimental options.
+	Experimental Experimental
+}
+
+type RequestOpts struct {
+	// Client is the http client used for all HTTP calls.
+	Client *http.Client
+	// HTTPContentType is the content type to use for the rpc request notification.
+	HTTPContentType string
+	// HTTPMethod is the HTTP method to use for the rpc request notification.
+	HTTPMethod string
+	// StaticHeaders are predefined headers that will be added to every request.
+	StaticHeaders http.Header
+	// TimestampFormat is the time format for the timestamp header.
+	TimestampFormat string
+	// TimestampHeader is the header name that should contain the timestamp. Example: X-BMCLIB-Timestamp
+	TimestampHeader string
+}
+
+type SignatureOpts struct {
+	// HeaderName is the header name that should contain the signature(s). Example: X-BMCLIB-Signature
+	HeaderName string
+	// AppendAlgoToHeaderDisabled decides whether to append the algorithm to the signature header or not.
+	// Example: X-BMCLIB-Signature becomes X-BMCLIB-Signature-256
+	// When set to true, a header will be added for each algorithm. Example: X-BMCLIB-Signature-256 and X-BMCLIB-Signature-512
+	AppendAlgoToHeaderDisabled bool
+	// IncludedPayloadHeaders are headers whose values will be included in the signature payload. Example: X-BMCLIB-My-Custom-Header
+	// All headers will be deduplicated.
+	IncludedPayloadHeaders []string
+}
+
+type HMACOpts struct {
+	// Hashes is a map of algorithms to a slice of hash.Hash (these are the hashed secrets). The slice is used to support multiple secrets.
+	Hashes map[Algorithm][]hash.Hash
+	// PrefixSigDisabled determines whether the algorithm will be prefixed to the signature. Example: sha256=abc123
+	PrefixSigDisabled bool
```

Comment:
Similar to other comments, whats the motivation to provide this level of flexibility?

Reply:
yeah, fair. the motivation is interoperability with systems where HMAC might have a specific requirement with regard to its payload signature header. I want to provide as much flexibility for users and bmclib will most likely be the new piece of the puzzle that will most likely have to be more flexible in order to interoperate properly.